render/detailed/node.go

```diff
@@ -0,0 +1,303 @@
+package detailed
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/weaveworks/scope/probe/docker"
+	"github.com/weaveworks/scope/probe/host"
+	"github.com/weaveworks/scope/probe/kubernetes"
+	"github.com/weaveworks/scope/probe/process"
+	"github.com/weaveworks/scope/render"
+	"github.com/weaveworks/scope/report"
+)
+
+// Node is the data type that's yielded to the JavaScript layer when
+// we want deep information about an individual node.
+type Node struct {
+	ID       string            `json:"id"`
+	Label    string            `json:"label"`
+	Rank     string            `json:"rank,omitempty"`
+	Pseudo   bool              `json:"pseudo,omitempty"`
+	Controls []ControlInstance `json:"controls"`
+	Metadata []MetadataRow     `json:"metadata,omitempty"`
+	Metrics  []MetricRow       `json:"metrics,omitempty"`
+	Children []ChildGroup      `json:"children,omitempty"`
+	Parents  []Parent          `json:"parents,omitempty"`
+}
+
+// ChildGroup is a topology-typed group of children for a Node.
+type ChildGroup struct {
+	Label      string         `json:"label"`
+	Nodes      []ChildSummary `json:"nodes"`
+	TopologyID string         `json:"topologyId"`
+}
+
+// ChildSummary is summary information about a child for a Node.
+type ChildSummary struct {
+	ID       string        `json:"id"`
+	Label    string        `json:"label"`
+	Linkable bool          `json:"linkable"` // Whether this node can be linked-to
+	Metadata []MetadataRow `json:"metadata,omitempty"`
+	Metrics  []MetricRow   `json:"metrics,omitempty"`
+}
+
+// Parent is the information needed to build a link to the parent of a Node.
+type Parent struct {
+	ID         string `json:"id"`
+	Label      string `json:"label"`
+	TopologyID string `json:"topologyId"`
+}
+
+// ControlInstance contains a control description, and all the info
+// needed to execute it.
+type ControlInstance struct {
+	ProbeID string `json:"probeId"`
+	NodeID  string `json:"nodeId"`
+	report.Control
+}
+
+// MakeNode transforms a renderable node to a detailed node. It uses
+// aggregate metadata, plus the set of origin node IDs, to produce tables.
+func MakeNode(r report.Report, n render.RenderableNode) Node {
+	var hosts, pods, containerImages, containers, applications []ChildSummary
+	for id, child := range n.Children {
+		if id == n.ID {
+			continue
+		}
+
+		switch child.Topology {
+		case "process":
+			applications = append(applications, processChildSummary(child))
+		case "container":
+			containers = append(containers, containerChildSummary(child))
+		case "container_image":
+			containerImages = append(containerImages, containerImageChildSummary(child))
+		case "pod":
+			pods = append(pods, podChildSummary(child))
+		case "host":
+			hosts = append(hosts, hostChildSummary(child))
+		}
```

Comment:
I'd like to see if we can remove the specialisation here - perhaps a function MakeChildSummary that uses NodeMetadata and NodeMetrics etc?